src/wallet/feebumper.cpp

```diff
@@ -57,6 +57,86 @@ static feebumper::Result PreconditionChecks(interfaces::Chain::Lock& locked_chai
+        errors.push_back(strprintf(
+            "New fee rate (%s) is lower than the minimum fee rate (%s) to get into the mempool -- ",
+            FormatMoney(newFeerate.GetFeePerK()),
+            FormatMoney(minMempoolFeeRate.GetFeePerK())));
+        return feebumper::Result::WALLET_ERROR;
+    }
+
+    CAmount new_total_fee = newFeerate.GetFee(maxTxSize);
+
+    CFeeRate incrementalRelayFee = std::max(wallet->chain().relayIncrementalFee(), CFeeRate(WALLET_INCREMENTAL_RELAY_FEE));
+
+    // Given old total fee and transaction size, calculate the old feeRate
+    CAmount old_fee = wtx.GetDebit(ISMINE_SPENDABLE) - wtx.tx->GetValueOut();
+    const int64_t txSize = GetVirtualTransactionSize(*(wtx.tx));
+    CFeeRate nOldFeeRate(old_fee, txSize);
+    // Min total fee is old fee + relay fee
+    CAmount minTotalFee = nOldFeeRate.GetFee(maxTxSize) + incrementalRelayFee.GetFee(maxTxSize);
+
+    if (new_total_fee < minTotalFee) {
+        errors.push_back(strprintf("Insufficient total fee %s, must be at least %s (oldFee %s + incrementalFee %s)",
+            FormatMoney(new_total_fee), FormatMoney(minTotalFee), FormatMoney(nOldFeeRate.GetFee(maxTxSize)), FormatMoney(incrementalRelayFee.GetFee(maxTxSize))));
+        return feebumper::Result::INVALID_PARAMETER;
+    }
+
+    CAmount requiredFee = GetRequiredFee(*wallet, maxTxSize);
+    if (new_total_fee < requiredFee) {
+        errors.push_back(strprintf("Insufficient total fee (cannot be less than required fee %s)",
+            FormatMoney(requiredFee)));
+        return feebumper::Result::INVALID_PARAMETER;
+    }
+
+    // Check that in all cases the new fee doesn't violate maxTxFee
+    const CAmount max_tx_fee = wallet->m_default_max_tx_fee;
+    if (new_total_fee > max_tx_fee) {
+        errors.push_back(strprintf("Specified or calculated fee %s is too high (cannot be higher than -maxtxfee %s)",
+                            FormatMoney(new_total_fee), FormatMoney(max_tx_fee)));
+        return feebumper::Result::WALLET_ERROR;
+    }
+
+    return feebumper::Result::OK;
+}
+
+static CFeeRate EstimateFeeRate(CWallet* wallet, const CWalletTx& wtx, CCoinControl& coin_control, CAmount& old_fee)
+{
+    // Get the fee rate of the original transaction. This is calculated from
+    // the tx fee/vsize, so it may have been rounded down. Add 1 satoshi to the
+    // result.
+    old_fee = wtx.GetDebit(ISMINE_SPENDABLE) - wtx.tx->GetValueOut();
+    int64_t txSize = GetVirtualTransactionSize(*(wtx.tx));
+    CFeeRate feerate(old_fee, txSize);
+    feerate += CFeeRate(1);
+
+    // The node has a configurable incremental relay fee. Increment the fee by
+    // the minimum of that and the wallet's conservative
+    // WALLET_INCREMENTAL_RELAY_FEE value to future proof against changes to
+    // network wide policy for incremental relay fee that our node may not be
+    // aware of. This ensures we're over the over the required relay fee rate
```

Comment:
"over the over the"
Though I see this typo exists in the old code as well so there's no need to fix that in this PR.